sql/src/main/java/org/apache/druid/sql/calcite/table/DatasourceTable.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.sql.calcite.table;
+
+import com.google.common.base.Preconditions;
+import org.apache.calcite.plan.RelOptTable;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.logical.LogicalTableScan;
+import org.apache.druid.query.DataSource;
+import org.apache.druid.query.TableDataSource;
+import org.apache.druid.segment.column.RowSignature;
+
+import java.util.Objects;
+
+public class DatasourceTable extends DruidTable
```

Comment:
the naming is going to get a bit confusing. We have `TableDataSource` and we have `DatasourceTable` 😄 
To me, `DataSource` is druid native/processing concept and `table` is druid SQL concept. But then we also have `TableDatasource` class in processing layer that has existed since ages. 

You could either rename it to `DatasourceSQLTable` so that there is less ambiguity between this class and `TableDataSource`. In any case, I think this class should have some javadocs for future readers who might trip over this naming ambiguity.

Reply:
Yeah, it is confusing. Didn't add "SQL" to the name since the class ultimately derives from a Calcite interface which is, by definition, a SQL concept. Also confusing because this class will be the place to combine physical and catalog data; but that catalog part is not yet part of the code. Did add Javadoc to explain what this class actually does. Check if that documentation addresses the confusion.